File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
		int left, right, sum = 0;
		List<List<Integer>> answer = new ArrayList<>();
		Arrays.sort(nums);

		for (int i = 0; i < nums.length - 2; i++) {
			if (i > 0 && nums[i] == nums[i - 1])
				continue;
			left = i + 1;
			right = nums.length - 1;
			while (left < right) {
				sum = nums[i] + nums[left] + nums[right];
				// 0 보다 작으면 left 이동하여 더 큰수 사용
				if (sum < 0) {
					left++;
				} else if (sum > 0) {
					right--;
				} else {
					// 0 인 경우
					answer.add(List.of(nums[i], nums[left], nums[right]));

					// left, right 의 중복값 건너뛰기
					while (left < right && nums[left] == nums[left + 1]) {
						left++;
					}
					while (left < right && nums[right] == nums[right - 1]) {
						right--;
					}

					// 정답이 나왔으므로 다음 left, right 사용
					left++;
					right--;
				}
			}

		}

		return answer;
    }
}

File: 0015-3sum/0015-3sum.kt
class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
    // 투 포인터를 활용해서 풀이
    var left: Int
    var right: Int
    var sum: Int

    // 정렬
    nums.sort()
    var answer: MutableList<List<Int>> = mutableListOf()

    for ((i, num) in nums.withIndex()) {
        if (i > 0 && nums[i] == nums[i - 1]) continue

        left = i + 1
        right = nums.size - 1
        while (left < right) {
            sum = num + nums[right] + nums[left]
            if (sum > 0) {
                right--
            } else if (sum < 0) {
                left++
            } else {
                // 0 이 되는 정답의 경우
                answer.add(listOf(num, nums[left], nums[right]))

                // left, right 에 중복 원소 넘어가기
                while (left < right && nums[left] == nums[left + 1]) {
                    left++
                }

                while (left < right && nums[right] == nums[right - 1]) {
                    right--
                }

                left++
                right--
            }
        }
    }

    return answer
    }
}